py_ecc/bls/ciphersuites.py

```diff
@@ -0,0 +1,187 @@
+    signature_to_G2,
+)
+
+
+class BaseG2Ciphersuite(ABC):
+    @abstractproperty
+    def DST(self):
+        pass
+
+    @staticmethod
+    def PrivToPub(privkey: int) -> BLSPubkey:
+        return G1_to_pubkey(multiply(G1, privkey))
+
+    @staticmethod
+    def KeyGen(IKM: bytes) -> Tuple[BLSPubkey, int]:
+        prk = hkdf_extract(b'BLS-SIG-KEYGEN-SALT-', IKM)
+        l = ceil((1.5 * ceil(log2(curve_order))) / 8)  # noqa: E741
+        okm = hkdf_expand(prk, b'', l)
+        x = big_endian_to_int(okm) % curve_order
+        return (BaseG2Ciphersuite.PrivToPub(x), x)
+
+    @staticmethod
+    def KeyValidate(PK: BLSPubkey) -> bool:
+        try:
+            pubkey_to_G1(PK)
+        except ValidationError:
+            return False
+        return True
+
+    @staticmethod
+    def _CoreSign(SK: int, message: bytes, DST: bytes) -> BLSSignature:
+        message_point = hash_to_G2(message, DST)
+        signature_point = multiply(message_point, SK)
+        return G2_to_signature(signature_point)
+
+    @staticmethod
+    def _CoreVerify(PK: BLSPubkey, message: bytes, signature: BLSSignature, DST: bytes) -> bool:
+        try:
+            signature_point = signature_to_G2(signature)
+            final_exponentiation = final_exponentiate(
+                pairing(
+                    signature_point,
+                    G1,
+                    final_exponentiate=False,
+                ) * pairing(
+                    hash_to_G2(message, DST),
+                    neg(pubkey_to_G1(PK)),
+                    final_exponentiate=False,
+                )
+            )
+            return final_exponentiation == FQ12.one()
+        except (ValidationError, ValueError, AssertionError):
+            return False
+
+    @staticmethod
+    def Aggregate(signatures: Iterable[BLSSignature]) -> BLSSignature:
+        accumulator = Z2  # Seed with the point at infinity
+        for signature in signatures:
+            signature_point = signature_to_G2(signature)
+            accumulator = add(accumulator, signature_point)
+        return G2_to_signature(accumulator)
+
+    @staticmethod
+    def _CoreAggregateVerify(pairs: Iterable[Tuple[BLSPubkey, bytes]],
+                             signature: BLSSignature, DST: bytes) -> bool:
+        try:
+            signature_point = signature_to_G2(signature)
+            accumulator = FQ12([1] + [0] * 11)
+            for pk, message in pairs:
+                pubkey_point = pubkey_to_G1(pk)
+                message_point = hash_to_G2(message, DST)
+                accumulator *= pairing(message_point, pubkey_point, final_exponentiate=False)
+            accumulator *= pairing(signature_point, neg(G1), final_exponentiate=False)
+            return final_exponentiate(accumulator) == FQ12.one()
+
+        except (ValidationError, ValueError, AssertionError):
+            return False
+
+    def Sign(self, SK: int, message: bytes) -> BLSSignature:
+        return self._CoreSign(SK, message, self.DST)
```

Comment:
> What I don't like about what I've done now is that `Sign` etc are not static methods because they need access to `self.DST`.  This means that it is necessary to instantiate the class before use which is ugly and a bad interface. Do you have a good idea on how to fix it?

I think you can:
1. Make `DST` a class variable.
2. Change the APIs which need `DST` to `classmethod`s.

```python
class BaseG2Ciphersuite(ABC):
    DST = b''

    @classmethod
    def Sign(cls, SK: int, message: bytes) -> BLSSignature:
        return self._CoreSign(SK, message, cls.DST)
```

Reply:
Thanks, works like a charm!